keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"int64,1,opt,name=observedGeneration"`
+    
+    // Installed is used by the Application Controller ot report the installed Components and Dependencies of  
+    // an Application.
+    Installed []string `json:"installed,omitempty" protobuf:"bytes,2,opt,name=installed"`
+    
+    // Ready is used to report the readiness of an Application. If an Application contains an ApplicationHealthCheck, 
+    // this field corresponds directly to the Readiness of the Pod generated from that health check. If the 
+    // Application does not contain an ApplicationHealthCheck, this field is set to Unknown.
+    Ready ConditionStatus json:`"ready,omitempty" protobuf:"bytes,3,opt,name=ready"`
+}
+
+// ApplicationHealthCheck contains a template for a Pod that performs an Application health check. The readiness 
+// of the Pod decides the readiness of the Application.
+type ApplicationHealthCheck struct {
+    
+    // Template is used to specify a Pod whose readiness will be used to determine the readiness of an Application.
+    Template v1.PodTemplateSpec
+}
+// ApplicationSpec is the specification object for an Application. It specifies the Application's components, 
+// dependencies, and an optional health check. Its selector can be used to select application components.
+type ApplicationSpec struct {
+    // Type is the type of the application (e.g. WordPress, MySQL, Cassandra).
+    Type string `json:"type" protobuf:"bytes,1,name=type"`
+    
+    // Components is a map of the applications components to the kinds created by the application (e.g. Pods, Services, 
+    // Deployments, CRDS, etc)
+    Components map [string]metav1.GroupVersionKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
+    
+    // HealthCheck is an optional application level health check.
+    HealthCheck* ApplicationHealthCheck `json:"healthCheck" protobuf:"bytes,5,opt,name=healthCheck"`
+    
+    // Version is an optional version indicator for the application.
+    Version string `json:"version,omitempty" protobuf:"bytes,6,opt,name=version"`
+    
+    // AboutURL is a URL pointing to any additional information about the application.
+    AboutURL string `json:"url,omitempty" protobuf:"bytes,7,opt,name=url"`
+    
+    // Description is a brief string description of the application.
+    Description string `json:"description,omitempty" protobuf:"bytes,8,opt,name=description"`
```

Comment:
In addition to this free-form description field, consider adding an (ordered?) list of label-value pairs, which a UI would render in tabular form to show the user "quick info" about the application. For example:
- Site / Admin URLs
- References to secrets that contain passwords
- Documentation links

This generic structure could replace some of the more contentious fields in the proposal (e.g. LicenseURL, DashboardURL).

To anticipate another objection -- we could not simply use annotations for this purpose, because there are many annotations that are not intended for human consumption. Also, I'd argue that making this an *ordered* list is important for showing the most salient info first.